src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -1331,3 +1336,117 @@ def fill_cache(self, job, force=False):
             self.logger.debug('cache fill is in progress.')
             return {'users': [], 'groups': []}
         return await self.middleware.call('cache.get', 'AD_cache')
+
+
+class WBStatusThread(threading.Thread):
+    def __init__(self, **kwargs):
+        super(WBStatusThread, self).__init__()
+        self.setDaemon(True)
+        self.middleware = kwargs.get('middleware')
+        self.logger = self.middleware.logger
+        self.finished = threading.Event()
+        self.state = msg.MSG_WINBIND_ONLINE
+
+    def parse_msg(self, data):
+        m = json.loads(data)
+        new_state = self.state
+        if not self.middleware.call_sync('activedirectory.config')['enable']:
+            self.logger.debug('Ignoring winbind message for disabled AD service: [%s]', m)
+            return
+
+        try:
+            new_state = DSStatus(m['winbind_message']).value
+        except Exception as e:
+            self.logger.debug('Received invalid winbind status message [%s]: %s', m, e)
+
+        if m['domain_name_netbios'] != self.middleware.call_sync('smb.config')['workgroup']:
+            self.logger.debug(
+                'Domain [%s] changed state to %s',
+                m['domain_name_netbios'],
+                DSStatus(m['winbind_message']).name
+            )
+            return
+
+        if self.state != new_state:
+            self.logger.debug(
+                'State of domain [%s] transistioned to [%s]',
+                m['forest_name'], DSStatus(m['winbind_message'])
+            )
+            self.middleware.call_sync('activedirectory._set_state', DSStatus(m['winbind_message']))
+            if new_state == DSStatus.FAULTED.value:
+                self.middleware.call_sync("alert.oneshot_create",
+                    "ActiveDirectoryDomainOffline", {"domain": m["domain_name_netbios"]}
+                )
+            else:
+                self.middleware.call_sync("alert.oneshot_delete",
+                    "ActiveDirectoryDomainOffline", {"domain": m["domain_name_netbios"]}
+                )
+
+        self.state = new_state
+
+    def read_messages(self):
+        while not self.finished.is_set():
+            with open('/var/run/samba4/.wb_fifo') as f:
+                while not self.finished.is_set():
+                    select.select([f], [], [f])
```

Comment:
Dont we need a timeout here? Otherwise it will block forever, even if `self.finished.is_set`